lib/Dialect/AIEVec/Transforms/VectorToAIEVecConversions.cpp

```diff
@@ -46,6 +46,49 @@ using namespace xilinx::aievec;
 // Utility functions
 //===----------------------------------------------------------------------===//
 
+static bool isNarrowingOp(Operation *op) {
+  if (isa<arith::TruncFOp>(op) || isa<arith::TruncIOp>(op))
+    return true;
+
+  if (auto srsOp = dyn_cast<aievec::SRSOp>(op)) {
+    auto srsOpSrcOp = srsOp.getSource().getDefiningOp();
+    if (isa<aievec::UPSOp>(srsOpSrcOp) || isa<aievec::CastOp>(srsOpSrcOp))
+      return true;
+  }
+  return false;
+}
+
+// Given a Value, if it is defined by a widening op (arith:ExtSIOp,
+// arith::ExtUIOp, arith::ExtFOp, aievec::UPSOp + aievec::SRSOp,
+// aievec::UPSOp + aievec::CastOp), return the source of the widening op.
+static std::optional<Value> getSourceOfWideningOp(Value src) {
+  if (auto extSIOp = src.getDefiningOp<arith::ExtSIOp>())
+    return extSIOp.getIn();
+  if (auto extUIOp = src.getDefiningOp<arith::ExtUIOp>())
+    return extUIOp.getIn();
+  if (auto extFOp = src.getDefiningOp<arith::ExtFOp>())
+    return extFOp.getIn();
+  if (auto srsOp = src.getDefiningOp<aievec::SRSOp>()) {
+    // Conversion through AIE intrinsics takes two steps:
+    //     1) Load to accumulator: aievec.ups
+    //     2) Move from accumulator: aievec.srs
+    auto srsSource = srsOp.getSource();
+    if (srsSource)
+      if (auto upsOp = srsSource.getDefiningOp<aievec::UPSOp>())
```

Comment:
Suppose that we have UPS from int16 => int32 followed by SRS from int32 to int8.
IMO, we should return sct::optional<Value>{} in this case.

Reply:
Indeed, that would be the thing to do since that's a narrowing operation, not a widening one. But we're not looking for `UPS` -> `SRS` sequences here, only `SRS` -> `UPS`.